compiler/src/dotty/tools/dotc/Driver.scala

```diff
@@ -31,35 +32,48 @@ class Driver {
 
   protected def emptyReporter: Reporter = new StoreReporter(null)
 
-  protected def doCompile(compiler: Compiler, fileNames: List[String])(using Context): Reporter =
-    if (fileNames.nonEmpty)
+  protected def doCompile(compiler: Compiler, fileNames: List[String])(using ctx: Context): Reporter =
+    if fileNames.nonEmpty then
       try
         val run = compiler.newRun
         run.compile(fileNames)
-
-        def finish(run: Run)(using Context): Unit =
-          run.printSummary()
-          if !ctx.reporter.errorsReported && run.suspendedUnits.nonEmpty then
-            val suspendedUnits = run.suspendedUnits.toList
-            if (ctx.settings.XprintSuspension.value)
-              report.echo(i"compiling suspended $suspendedUnits%, %")
-            val run1 = compiler.newRun
-            for unit <- suspendedUnits do unit.suspended = false
-            run1.compileUnits(suspendedUnits)
-            finish(run1)(using MacroClassLoader.init(ctx.fresh))
-
-        finish(run)
+        finish(compiler, run)
       catch
-        case ex: FatalError  =>
+        case ex: FatalError =>
           report.error(ex.getMessage) // signals that we should fail compilation.
         case ex: TypeError =>
           println(s"${ex.toMessage} while compiling ${fileNames.mkString(", ")}")
-          throw ex
         case ex: Throwable =>
           println(s"$ex while compiling ${fileNames.mkString(", ")}")
           throw ex
     ctx.reporter
-  end doCompile
+
+  protected def doCompileFiles(compiler: Compiler,  files: List[AbstractFile])(using Context): Reporter =
```

Comment:
Can we avoid the code duplication between doCompile and doCompileFiles?

Reply:
This is my best attempt so far: https://github.com/lampepfl/dotty/pull/10607/commits/168a4162379f271eb95c385f9f898e6547f4635c
Let me know what you think!